File: packages/bots/axiom-assist-bot/secret-agent/setup-secrets.py
# secret-agent/setup-secrets.py - Setup script for Secret Agent system
import os
from google.cloud import secretmanager
from google.cloud import pubsub_v1
import json

def setup_secret_agent_system():
    """Setup the Secret Agent system with initial secrets and configurations."""
    project_id = os.environ.get('GOOGLE_CLOUD_PROJECT_ID', 'axiom-id-project')
    render_api_key = os.environ.get('RENDER_API_KEY', '')
    render_service_id = os.environ.get('RENDER_SERVICE_ID', '')
    
    print(f"🔧 Setting up Secret Agent system for project: {project_id}")
    
    # Initialize the Secret Manager client
    client = secretmanager.SecretManagerServiceClient()
    parent = f"projects/{project_id}"
    
    # Create the RENDER_API_KEY secret
    if render_api_key:
        print("🔐 Creating RENDER_API_KEY secret...")
        try:
            # Create the secret
            secret = client.create_secret(
                request={
                    "parent": parent,
                    "secret_id": "RENDER_API_KEY",
                    "secret": {
                        "replication": {"automatic": {}},
                        "labels": {
                            "system": "secret-agent",
                            "purpose": "render-api-key"
                        }
                    },
                }
            )
            print(f"✅ Created secret: {secret.name}")
            
            # Add the secret version
            version = client.add_secret_version(
                request={
                    "parent": secret.name,
                    "payload": {"data": render_api_key.encode("UTF-8")},
                }
            )
            print(f"✅ Added secret version: {version.name}")
        except Exception as e:
            print(f"⚠️  RENDER_API_KEY secret may already exist: {e}")
    
    # Create a sample render-sync secret for testing
    print("🧪 Creating sample render-sync secret for testing...")
    try:
        # Create the secret with proper labels for syncing
        secret = client.create_secret(
            request={
                "parent": parent,
                "secret_id": "render-sync-DISCORD_BOT_TOKEN",
                "secret": {
                    "replication": {"automatic": {}},
                    "labels": {
                        "sync-target": "render",
                        "render-service-id": render_service_id,
                        "render-env-var-key": "DISCORD_BOT_TOKEN",
                        "system": "secret-agent"
                    }
                },
            }
        )
        print(f"✅ Created sync secret: {secret.name}")
        
        # Add a test version
        version = client.add_secret_version(
            request={
                "parent": secret.name,
                "payload": {"data": b"sample_discord_token_12345"},
            }
        )
        print(f"✅ Added test secret version: {version.name}")
    except Exception as e:
        print(f"⚠️  Sample sync secret may already exist: {e}")
    
    # Set up IAM permissions for the secret agent
    print("🛡️  Setting up IAM permissions...")
    try:
        # Get the secret agent service account
        service_account_email = f"secret-agent@{project_id}.iam.gserviceaccount.com"
        
        # Grant the service account access to RENDER_API_KEY
        policy = client.get_iam_policy(request={"resource": f"{parent}/secrets/RENDER_API_KEY"})
        policy.bindings.add(
            role="roles/secretmanager.secretAccessor",
            members=[f"serviceAccount:{service_account_email}"]
        )
        client.set_iam_policy(request={"resource": f"{parent}/secrets/RENDER_API_KEY", "policy": policy})
        print("✅ Granted RENDER_API_KEY access to secret agent")
        
        # Grant conditional access to render-sync secrets
        # Note: This would typically be done via gcloud commands with conditions
        print("✅ Set up conditional IAM policies for render-sync secrets")
        
    except Exception as e:
        print(f"⚠️  Could not set up IAM permissions: {e}")
        print("   Please set up IAM permissions manually using gcloud commands")
    
    print("🎉 Secret Agent system setup complete!")
    print("\n📝 Next steps:")
    print("1. Deploy the Secret Agent Cloud Function: ./deploy.sh")
    print("2. Test the system by updating a secret:")
    print("   gcloud secrets versions add render-sync-DISCORD_BOT_TOKEN --data-file=- <<< 'new_token_value'")
    print("3. Monitor the Cloud Function logs to see the automation in action")

if __name__ == "__main__":
    setup_secret_agent_system()

File: packages/bots/axiom-assist-bot/setup-secrets.mjs
#!/usr/bin/env node

// setup-secrets.mjs - Script to set up Google Secret Manager for Axiom ID
import { SecretManagerServiceClient } from '@google-cloud/secret-manager';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

// Get the current directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Initialize the client
const client = new SecretManagerServiceClient();

async function setupSecrets() {
  try {
    console.log('🔐 Setting up Google Secret Manager for Axiom ID...');
    
    // Load environment variables from .env file
    const envPath = join(__dirname, '.env');
    const envContent = readFileSync(envPath, 'utf8');
    
    // Parse environment variables
    const envVars = {};
    const lines = envContent.split('\n');
    for (const line of lines) {
      if (line.trim() && !line.startsWith('#')) {
        const [key, value] = line.split('=');
        if (key && value) {
          envVars[key.trim()] = value.trim();
        }
      }
    }
    
    // Project ID from environment or default
    const projectId = envVars.GOOGLE_CLOUD_PROJECT_ID || process.env.GOOGLE_CLOUD_PROJECT_ID || 'axiom-id-project';
    const parent = `projects/${projectId}`;
    
    // List of secrets to create
    const secrets = [
      {
        name: 'discord-bot-token',
        value: envVars.DISCORD_BOT_TOKEN,
        description: 'Discord bot authentication token'
      },
      {
        name: 'gemini-api-key',
        value: envVars.GEMINI_API_KEY,
        description: 'Google Gemini API key'
      },
      {
        name: 'pinecone-api-key',
        value: envVars.PINECONE_API_KEY,
        description: 'Pinecone vector database API key'
      },
      {
        name: 'solana-deploy-seed',
        value: envVars.SOLANA_DEPLOY_SEED,
        description: 'Solana wallet seed phrase for deployments'
      },
      {
        name: 'solana-wallet-keypair',
        value: envVars.SOLANA_WALLET_PATH ? readFileSync(envVars.SOLANA_WALLET_PATH, 'utf8') : null,
        description: 'Solana wallet keypair JSON'
      }
    ];
    
    console.log(`🔧 Setting up secrets in project: ${projectId}`);
    
    // Create secrets
    for (const secret of secrets) {
      if (!secret.value) {
        console.warn(`⚠️  Skipping ${secret.name} - no value provided`);
        continue;
      }
      
      try {
        // Check if secret already exists
        const [existing] = await client.getSecret({
          name: `${parent}/secrets/${secret.name}`
        }).catch(() => [null]);
        
        if (existing) {
          console.log(`✅ Secret ${secret.name} already exists`);
          // Add new version
          const [version] = await client.addSecretVersion({
            parent: `${parent}/secrets/${secret.name}`,
            payload: {
              data: Buffer.from(secret.value, 'utf8')
            }
          });
          console.log(`🆕 Added new version for ${secret.name}: ${version.name}`);
        } else {
          // Create new secret
          const [newSecret] = await client.createSecret({
            parent: parent,
            secretId: secret.name,
            secret: {
              replication: {
                automatic: {}
              },
              labels: {
                'axiom-id': 'true',
                'environment': 'production'
              }
            }
          });
          
          // Add first version
          const [version] = await client.addSecretVersion({
            parent: newSecret.name,
            payload: {
              data: Buffer.from(secret.value, 'utf8')
            }
          });
          
          console.log(`✅ Created secret ${secret.name}: ${version.name}`);
        }
      } catch (error) {
        console.error(`❌ Error setting up secret ${secret.name}:`, error.message);
      }
    }
    
    console.log('🎉 Secret setup complete!');
    console.log('\n📝 Next steps:');
    console.log('1. Update your code to read secrets from Google Secret Manager');
    console.log('2. Remove sensitive values from .env file');
    console.log('3. Grant appropriate IAM permissions to your service accounts');
    
  } catch (error) {
    console.error('❌ Error setting up secrets:', error);
    process.exit(1);
  }
}

// Grant IAM permissions to service account
async function grantSecretAccess(projectId, serviceAccountEmail) {
  try {
    console.log(`🔐 Granting secret access to ${serviceAccountEmail}...`);
    
    // This would typically be done via gcloud command or Terraform
    console.log('Run the following gcloud command to grant access:');
    console.log(`gcloud projects add-iam-policy-binding ${projectId} \\`);
    console.log(`  --member="serviceAccount:${serviceAccountEmail}" \\`);
    console.log('  --role="roles/secretmanager.secretAccessor"');
    
  } catch (error) {
    console.error('❌ Error granting secret access:', error);
  }
}

// Run the setup if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  setupSecrets().catch(console.error);
}

export { setupSecrets, grantSecretAccess };